src/logql-query.ts

```diff
@@ -1,48 +1,237 @@
+    labelsInFilter: Array<LabelMatcher> | undefined;
+
+  syntaxTree.iterate({
+    from: node.from,
+    to: node.to,
+    enter(selectorsNode) {
+      if (selectorsNode.name === 'Pipe') {
+        operator = query.slice(selectorsNode.from, selectorsNode.to);
+
+        return false;
+      } else if (
+        ['PipeExact', 'PipeMatch', 'Nre', 'Neq'].includes(selectorsNode.name)
+      ) {
+        operator = query.slice(selectorsNode.from, selectorsNode.to);
+
+        return false;
+      } else if (
+        [
+          'LabelParser',
+          'JsonExpressionParser',
+          'LabelFilter',
+          'LineFormatExpr',
+          'LabelFormatExpr',
+          'String',
+        ].includes(selectorsNode.name)
+      ) {
+        value = query.slice(selectorsNode.from, selectorsNode.to);
+
+        if (selectorsNode.name === 'LabelFilter') {
+          labelsInFilter = parseMatchers(syntaxTree, selectorsNode, query);
+        }
+
+        return false;
+      }
+    },
+    leave(selectorsNode) {
+      if (selectorsNode.name === 'PipelineStage') {
+        pipelineStages.push({ operator, value, labelsInFilter });
+        labelsInFilter = undefined;
+      }
+    },
+  });
+
+  return pipelineStages;
+};
 
 export class LogQLQuery {
-  selectors: Array<string> = [];
-  pipeline: Array<string> = [];
+  streamSelector: Array<LabelMatcher> = [];
+  pipeline: Array<PipelineStage> = [];
 
   constructor(query: string) {
-    const parts = query.match(/.*{([^}]*)}(.*)/);
-    this.selectors =
-      parts && parts[1].length > 0
-        ? parts[1].split(',').filter(notEmptyString).map(trim)
-        : [];
-    this.pipeline =
-      parts && parts[2].length > 0
-        ? parts[2].split('|').filter(notEmptyString).map(trim)
-        : [];
+    const syntaxTree = parser.parse(query);
+    let parsedMatchers: Array<LabelMatcher> = [];
+    let parsedPipeline: Array<PipelineStage> = [];
+
+    syntaxTree.iterate({
+      enter(node) {
+        if (node.name === 'Selector' && parsedMatchers.length === 0) {
+          parsedMatchers = parseMatchers(syntaxTree, node, query);
+
+          return false;
+        } else if (
+          node.name === 'PipelineExpr' &&
+          parsedPipeline.length === 0
+        ) {
+          parsedPipeline = parsePipelineStages(syntaxTree, node, query);
+
+          return false;
+        }
+      },
+    });
+
+    this.streamSelector = parsedMatchers;
+    this.pipeline = parsedPipeline;
   }
 
-  appendSelector = (newSelectors: string | Array<string>) => {
-    if (Array.isArray(newSelectors)) {
-      this.selectors.push(...newSelectors);
-    } else {
-      this.selectors.push(newSelectors);
+  removeSelectorMatcher = (matcherToRemove: Partial<LabelMatcher>) => {
+    this.streamSelector = this.streamSelector.filter((matcher) => {
+      let include = true;
+
+      if (matcherToRemove.label) {
+        include = include && matcher.label !== matcherToRemove.label;
+      }
+
+      return include;
+    });
```

Comment:
How about (not tested)?

```suggestion
    this.streamSelector = this.streamSelector.filter(
      (matcher) => matcherToRemove.label && matcher.label !== matcherToRemove.label,
    );
```